fpdf/drawing.py

```diff
@@ -0,0 +1,4207 @@
+from typing import Optional, NamedTuple, Union
+
+__pdoc__ = {"force_nodocument": False}
+
+
+def force_nodocument(item):
+    """A decorator that forces pdoc not to document the decorated item (class or method)"""
+    __pdoc__[item.__qualname__] = False
+    return item
+
+
+@force_nodocument
+def force_document(item):
+    """A decorator that forces pdoc to document the decorated item (class or method)"""
+    __pdoc__[item.__qualname__] = True
+    return item
+
+
+# type alias:
+Number = Union[int, float, decimal.Decimal]
+NumberClass = (int, float, decimal.Decimal)
+
+
+# this maybe should live in fpdf.syntax
+class Raw(str):
+    """str subclass signifying raw data to be directly emitted to PDF without transformation."""
+
+
+class Name(str):
+    """str subclass signifying a PDF name, which are emitted differently than normal strings."""
+
+
+WHITESPACE = frozenset("\0\t\n\f\r ")
+"""Characters PDF considers to be whitespace."""
+EOL_CHARS = frozenset("\n\r")
+"""Characters PDF considers to mark the end of a line."""
+DELIMITERS = frozenset("()<>[]{}/%")
+"""Special delimiter characters"""
+
+
+NAME_ESC = re.compile(
+    b"[^" + bytes(v for v in range(33, 127) if v not in b"()<>[]{}/%#\\") + b"]"
+)
+STR_ESC = re.compile(r"[\n\r\t\b\f()\\]")
+STR_ESC_MAP = {
+    "\n": r"\n",
+    "\r": r"\r",
+    "\t": r"\t",
+    "\b": r"\b",
+    "\f": r"\f",
+    "(": r"\(",
+    ")": r"\)",
+    "\\": r"\\",
+}
+
+
+class GraphicsStateDictRegistry(OrderedDict):
+    """
+    A container providing deduplication of graphics state dictionaries across a PDF.
+    """
+
+    def register_style(self, style):
+        sdict = style.to_pdf_dict()
+
+        # empty style does not need a dictionary
+        if not sdict:
+            return None
+
+        try:
+            return self[sdict]
+        except KeyError:
+            pass
+
+        name = Name(f"GS{len(self)}")
+        self[sdict] = name
+
+        return name
+
+
+def _check_range(value, minimum=0.0, maximum=1.0):
+    if not minimum <= value <= maximum:
+        raise ValueError(f"{value} not in range [{minimum}, {maximum}]")
+
+    return value
+
+
+def number_to_str(number):
+    """
+    Convert a decimal number to a minimal string representation (no trailing 0 or .).
+
+    Args:
+        number (Number): the number to be converted to a string.
+
+    Returns:
+        The number's string representation.
+    """
+    # this approach tries to produce minimal representations of floating point numbers
+    # but can also produce "-0".
+    return f"{number:.4f}".rstrip("0").rstrip(".")
+
+
+# this maybe should live in fpdf.syntax
+def render_pdf_primitive(primitive):
+    """
+    Render a Python value as a PDF primitive type.
+
+    Container types (tuples/lists and dicts) are rendered recursively. This supports
+    values of the type Name, str, bytes, numbers, booleans, list/tuple, and dict.
+
+    Any custom type can be passed in as long as it provides a `pdf_repr` method that
```

Comment:
yes, I agree that there are a few partial implementations of this throughout the codebase. the `FPDF` class also does a lot of manual dictionary/string construction in various locations, but if it ain't broke, don't fix them.

I implemented this automatic serializer to simplify the serialization of various drawing-related objects (largely the style dictionary), since none of the existing implementations handled recursively converting a data structure (e.g. a dictionary with a list or a Name as a value). To me, it definitely makes sense for this to live in the `fpdf.syntax` module and be used elsewhere, but, for better or worse, I was trying to avoid touching a bunch of other parts of the codebase when making these changes (this  PR is already big enough as is, I think).